src/main/java/io/oneko/project/rest/ProjectController.java

```diff
@@ -139,33 +141,35 @@ ProjectExportDTO exportProject(@PathVariable UUID id) {
 
 	@PreAuthorize("hasAnyRole('ADMIN', 'DOER', 'VIEWER')")
 	@PostMapping("/{id}/version/{versionId}/deploy")
-	ProjectDTO triggerDeploymentOfVersion(@PathVariable UUID id, @PathVariable UUID versionId) {
+	@ResponseStatus(HttpStatus.NO_CONTENT)
+	void triggerDeploymentOfVersion(@PathVariable UUID id, @PathVariable UUID versionId) {
 		WritableProject project = getProjectOr404(id).writable();
 		WritableProjectVersion projectVersion = project.getVersionById(versionId)
 				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Project version with id " + versionId + " not found"));
-		final ReadableProjectVersion deployedVersion = deploymentManager.deploy(projectVersion);
-		return dtoMapper.projectToDTO(deployedVersion.getProject());
+		final var deployTask = deploymentManager.deployAsync(projectVersion);
+		handleImmediateAsyncExceptionWithoutBlocking(deployTask);
 	}
 
-	@PreAuthorize("hasAnyRole('ADMIN', 'DOER', 'VIEWER')")
-	@PostMapping("/deploy/url")
-	ProjectDTO triggerDeploymentOfVersionWithUrl(@RequestBody String url) {
-		final var projectAndVersion = projectRepository.getByDeploymentUrl(url)
-				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "No version with url " + url + " found."));
-
-		final UUID versionId = projectAndVersion.getRight().getUuid();
-		final WritableProjectVersion version = projectAndVersion.getLeft().writable().getVersionById(versionId).orElseThrow();
-		final ReadableProjectVersion deployedVersion = deploymentManager.deploy(version);
-		return dtoMapper.projectToDTO(deployedVersion.getProject());
-	}
 
 	@PreAuthorize("hasAnyRole('ADMIN', 'DOER', 'VIEWER')")
 	@PostMapping("/{id}/version/{versionId}/stop")
 	void stopDeployment(@PathVariable UUID id, @PathVariable UUID versionId) {
 		WritableProject project = getProjectOr404(id).writable();
 		WritableProjectVersion projectVersion = project.getVersionById(versionId)
 				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Project version with id " + versionId + " not found"));
-		deploymentManager.stopDeployment(projectVersion);
+		final var stopTask = deploymentManager.stopDeploymentAsync(projectVersion);
+		handleImmediateAsyncExceptionWithoutBlocking(stopTask);
+	}
+
+	private <T> void handleImmediateAsyncExceptionWithoutBlocking(CompletableFuture<T> future) {
+		if (future.isCompletedExceptionally()) {
```

Comment:
If triggering the deployment fails already during initialization we might as well throw an exception synchronously.
This way we might get rid of this error-handling here, which to me look's out of place.

Reply:
Changed.